File: Design and Verification Training Program/Veriest/Scripting/Perl/calc.pl
#!/usr/bin/perl -w
use strict;
use Data::Dumper;

my $math_expr = $ARGV[0];;
#my $math_expr = "12+5*77-30*2+4-9";
my @math_expr_split_mult;
@math_expr_split_mult = split(/[*]+/, $math_expr);
#print "@math_expr_split_mult \n";
my $index;
for ($index = 1; $index <= $#math_expr_split_mult; $index = $index + 2){
    splice @math_expr_split_mult, $index, 0, '*'; # insert "*" as array element at the split points of the string
}
#print "@math_expr_split_mult \n";

my $item;
my @math_expr_split_plus;
foreach $item (@math_expr_split_mult){
    push @math_expr_split_plus, split(/[+]+/, $item);
}
#print "@math_expr_split_plus \n";

my @minus_arr;
my @temp;
my @math_expr_split_minus;
foreach $item (@math_expr_split_plus){
    if ($item =~ m/-/){ # if item contains "-"
        @temp = split(/[-]+/, $item);
        #print("@temp \n");
        @minus_arr = ($temp[0], -$temp[1]); # temp array for minuend, and subtrahend with minus sign
        #print "@minus_arr\n";
        push (@math_expr_split_minus, @minus_arr); # add the minus_arr current items to the end of math_expr_split_minus array
    } else {
        push (@math_expr_split_minus, $item); # add the current item to the end of math_expr_split_minus array
    }
}
#print "@math_expr_split_minus \n";

my @mult;
my @add;
for ($index = 0; $index <= $#math_expr_split_minus; $index++){
    if ($math_expr_split_minus[$index] eq '*'){
         push (@mult, $math_expr_split_minus[$index-1], $math_expr_split_minus[$index+1]); # add the previous and next items to the end of mult array
    } elsif (($math_expr_split_minus[$index-1] ne '*') && ($math_expr_split_minus[$index+1] ne '*')){
        push (@add, $math_expr_split_minus[$index]); # add the current item to the end of add array
    }
}
#print "@mult \n@add \n";
# create hash with keys: '*' and '+':
my %calc = ();
$calc{'*'} = [@mult];
$calc{'+'} = [@add];
#print Dumper(\%calc);

my $little_mult;
my @little_mults;
my $key;
foreach $key (keys %calc)
{
  if ($key eq '*'){
      for ($index = 0; $index < scalar(@{$calc{$key}}); $index = $index + 2){
          $little_mult =  $calc{$key}[$index] * $calc{$key}[$index + 1];
          push (@little_mults, $little_mult); # add the current multiplication result to the end of little_mults array
      }
  }
}
#print "@little_mults \n";
push ((@{$calc{'+'}}), @little_mults); # add multiplications results to value of '+' key
#print Dumper(\%calc);

my $final_result = 0;
foreach $key (keys %calc)
{
  if ($key eq '+'){
      for ($index = 0; $index < scalar(@{$calc{$key}}); $index++){
          $final_result=  $final_result + $calc{$key}[$index]; # calculate final result
      }
  }
}
print "$final_result \n"

File: Design and Verification Training Program/Veriest/Scripting/Python/calc.py

#!/usr/bin/python
import sys

#str = "12+5*77-30*2+4-9"
#str = input("Please insert your math expression\n")
#print(eval(str))

# get user input from terminal
str = sys.argv[1]
# split expression on *
new = str.split("*")
print(new)
for i in range (1,(len(new)+1),2):
    new.insert(i,'*')
print(new)
# split the sub-expressions on +
plus = []
for j in new:
    plus.append(j.split("+"))
print(plus)
for k in plus:
    for l in range (1,(len(k)),2):
        k.insert(l,'+')
print(plus)
# split the sub-expressions on -
minus = []
for s in plus:
    for x in s:
        minus.append(x.split("-"))
print(minus)
for t in minus:
    for q in range (1,(len(t)),2):
        t.insert(q,'-')
print(minus)
# flatten the last list to be list of strings
flat_list = []
for sublist in minus:
    for item in sublist:
        flat_list.append(item)
print(flat_list)
# calculate the multiplications
for element in flat_list:
    if element == "*":
        indx = flat_list.index(element)
        mult = int(flat_list[indx-1]) * int(flat_list[indx+1])
        flat_list = flat_list[:indx-1] + flat_list[indx+2 :]
        flat_list.insert(indx-1, mult)
        print(flat_list)
# calculate the subtractions and additions    
for lmnt in flat_list:
    if lmnt == "+":
        indx = flat_list.index(lmnt)
        add = int(flat_list[indx-1]) + int(flat_list[indx+1])
        flat_list = flat_list[:indx-1] + flat_list[indx+2 :]
        flat_list.insert(indx-1, add)
        print(flat_list)
    elif lmnt == "-":
        indx = flat_list.index(lmnt)
        sub = int(flat_list[indx-1]) - int(flat_list[indx+1])
        flat_list = flat_list[:indx-1] + flat_list[indx+2 :]
        flat_list.insert(indx-1, sub)
        print(flat_list)
        
#print(*flat_list)
print(flat_list[0])

